Mdd psql user aws fix

michael-dev2rights opened the pull request:
##### SUMMARY

This is an attempt to fix Ansible Bug #18933 - make postgresql user creation work on Amazon RDS instances.  It has worked on our internal branch based off Ansible stable 2.3 and has been cherry picked to devel for upstream inclusion into the project.  

##### ISSUE TYPE

 - Bugfix Pull Request


##### COMPONENT NAME

posgresql_user module 

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/mikedd/dev/ansible/ansible-add-modules/library']
  ansible python module location = /home/mikedd/dev/ansible/ansible-add-modules/lib/ansible
  executable location = /home/mikedd/dev/ansible/ansible-add-modules/bin/ansible
  python version = 2.7.12 (default, Jul  1 2016, 15:12:24) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

Not yet nearly fully tested.

mikedlr committed:
postgresql_user module - transaction logic hacks to allow recovery from failed select

mikedlr committed:
postgresql_user - PEP8 and style fixes to make debugging easier

mikedlr committed:
postgresql_user - move password changing logic to separate function

mikedlr committed:
postgresql_user - trap failure in case where there is no access to pg_authid

mikedlr committed:
postgresql_user - further PEP8 fixes

gundalow commented:
You wrote:
> Not yet nearly fully tested.
What other testing needs to be done?

michael-dev2rights commented:
@gundalow

1) someone should run the existing integration tests; probably after ensuring that they include user creation by this module and database changes via another module using that user.  
2) preferably, someone should write an integration test case which 
* creates an Amazon RDS with a database set up on it;  
* creates a user on that database via this module
* does something to the database
* changes the user's password
* does something to the database with the new password (also fails with old??)
* repeats the operation that changed the password with the same password (expected state on RDS is changed even though password won't actualy have changed)
* does something to the database

michael-dev2rights commented:
Beyond that it would be nice to have unit test cases but the refactoring I started should be continued with extreme prejudice, including converting all SQL string statements to prepared statements.

nerzhul commented:
Hello,
Warning if you refactor i have a pending PR on the module which adds IN_ROLE support (i use it at work to create nested developers groups on schemas)

nerzhul commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -267,46 +278,84 @@ def user_add(cursor, user, password, role_attr_flags, encrypted, expires):
+            # different
+            return True
+
+    if password == current_role_attrs['rolpassword']:
```
just return password != current_role_attrs['rolpassword']

nerzhul commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -267,46 +278,84 @@ def user_add(cursor, user, password, role_attr_flags, encrypted, expires):
+    if encrypted == 'ENCRYPTED' and not password.startswith('md5'):
+        try:
+            from passlib.hash import postgres_md5 as pm
+            if pm.encrypt(password, user) == current_role_attrs['rolpassword']:
```
just return pm.encrypt(password, user) != current_role_attrs['rolpassword']:

nerzhul commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -267,46 +278,84 @@ def user_add(cursor, user, password, role_attr_flags, encrypted, expires):
+
+    if encrypted == 'ENCRYPTED' and not password.startswith('md5'):
+        try:
+            from passlib.hash import postgres_md5 as pm
```
please don't import here, import globally

nerzhul commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -267,46 +278,84 @@ def user_add(cursor, user, password, role_attr_flags, encrypted, expires):
+    # Do we actually need to do anything?
+
+    if encrypted == 'ENCRYPTED' and password.startswith('md5'):
+        if password == current_role_attrs['rolpassword']:
```
same as others, return directly test

pilou- commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -669,7 +754,10 @@ def main():
+    # apparently port is automagically handled below;  comment out to relieve
+    # flake8 error; delete when tests verify
+    #
+    # port = module.params["port"]
```
Remove the whole unused line (and the comments).

pilou- reviewed (commented):
Please, update the description of 'no_password_changes'.

mikedlr committed:
postgresql_user - Simplify password change logic and improve imports according to suggestions from PR review

pilou- commented:
With the recent changes of the postgresql_user module, I am not sure these changes are needed anymore. Moreover the documentation states that `no_password_changes` parameter must be enabled when working with AWS RDS.

mikedlr committed:
postgresql_user - Eliminate pep8/blank line errors introduced in merge

michael-dev2rights commented:
@pilou; small comment: there is one way  to check if a password needs updated which is to use it.  If it works then it's already updated.  I don't know if this is worth the cost compared to just updating no matter what but it's a potential option.

As far the no_password_change goes; how recent is it?  This module is marked as stable_interface so it's quite important to maintain backward compatibility.

Apart from that I agree with your comments.

michael-dev2rights commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -273,41 +286,72 @@ def user_add(cursor, user, password, role_attr_flags, encrypted, expires):
-            elif encrypted == 'ENCRYPTED':
-                hashed_password = 'md5{0}'.format(md5(to_bytes(password) + to_bytes(user)).hexdigest())
-                if hashed_password != current_role_attrs['rolpassword']:
-                    pwchanging = True
```
discussion with @pilou- move this part into the "should we change password" function above

abadger commented on lib/ansible/modules/database/postgresql/postgresql_user.py:
```diff
@@ -220,14 +223,22 @@
+except ImportError:
+    passlib_hash_found = False
+else:
+    passlib_hash_found = True
```
Just a couple style note, usually we import from stdlib, then from 3rd party libs, and then from ansible libs.

For setting whether passlib is found, you can do it like this to save a line:
``` python
try:
    from passlib.hash import postgres_md5 as pm
    passlib_hash_found = True
except ImportError:
    passlib_hash_found = False
```

Since passlib_hash_found is a toplevel constant, style-wise it should be all uppercase:
PASSLIB_HASH_FOUND

abadger commented:
I found some style things but nothing important.

@nerzhul and @pilou- Does this change look good to you now?  I'll merge when you think it's ready.  

Also I see that https://github.com/ansible/ansible/pull/22613 currently needs to be rebased.  Perhaps after this goes in, @nerzhul would rebase that and then @michael-dev2rights would care to review it.  And then, when it's ready, one of you can ping me to merge that one too?

pilou- commented:
@abadger I discussed about this PR with @michael-dev2rights during ansible fest ([see](https://github.com/ansible/ansible/pull/23988#discussion_r123284115)), there is a problem (due to a complex rebase): `passlib.hash import postgres_md5` must not be reintroduced, `passlib.hash import postgres_md5` is sufficient.

pilou- committed:
Don't reintroduce passlib, remove useless query

pilou- committed:
Check behaviour when pg_authid relation isn't readable

TASK [postgresql : Normal user isn't allowed to access pg_authid relation:
      password comparison will fail, password will be updated] ***
An exception occurred during task execution. To see the full traceback,
use -vvv. The error was: psycopg2.ProgrammingError: permission denied
for relation pg_authid

pilou- commented:
I added 2 commits (on a [branch](https://github.com/pilou-/ansible/commits/mdd-psql-user-aws-fix) based on this one):
- the first commit (https://github.com/pilou-/ansible/commit/67193cb2d97c9273e5b7d851521a7a56a8229fba) adds a test, I checked the test fails on devel and succeed on this branch
- the second one (https://github.com/pilou-/ansible/commit/ec27231cc00c9df43c6ea14dda78098ec4ee8687) contains some fixes

@abadger could you update this pull-request adding these two commits and then merge ?

michael-dev2rights commented:
@pilou-  I merged your changes into this branch.  Thanks.

@abadger I think we should merge this PR soon since it's a proper bug fix and now includes useful tests.  

@nerzhul any objections?

abadger commented:
I've merged to devel (for 2.4.0).  If @nerzhul pokes me on IRC we can decide whether to backport to 2.3.x as well.
